addons/account/models/account_account.py

```diff
@@ -255,14 +259,19 @@ def _check_journal_consistency(self):
                 account=account.display_name
             ))
 
-    @api.constrains('company_id')
+    @api.constrains('company_ids')
     def _check_company_consistency(self):
-        for company, accounts in tools.groupby(self, lambda account: account.company_id):
-            if self.env['account.move.line'].search_count([
+        if accounts_without_company := self.filtered(lambda a: not a.sudo().company_ids):
+            raise ValidationError(
+                _("The following accounts must be assigned to at least one company:")
+                + "\n" + "\n".join(f"- {account.display_name}" for account in accounts_without_company)
+            )
+        for companies, accounts in tools.groupby(self, lambda a: a.company_ids):
```

Comment:
Better time complexity, and you can do `accounts.ids` below too
```suggestion
        for companies, accounts in self.grouped(lambda a: a.company_ids).items():
```